src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -1439,23 +1429,134 @@ public static IEnumerable<int> Range(int start, int count)
         {
             long max = ((long)start) + count - 1;
             if (count < 0 || max > Int32.MaxValue) throw Error.ArgumentOutOfRange("count");
-            return RangeIterator(start, count);
+            return new RangeIterator(start, count);
         }
 
-        private static IEnumerable<int> RangeIterator(int start, int count)
+        private sealed class RangeIterator : Iterator<int>, IArrayProvider<int>
         {
-            for (int end = start + count; start != end; start++) yield return start;
+            private readonly int _start;
+            private readonly int _end;
+            
+            public RangeIterator(int start, int count)
+            {
+                _start = start;
+                _end = start + count;
+            }
+            
```

Comment:
#3064 I removed all the trailing whitespace throughout because why not, and that nit with the MoveNext too.